nemo/collections/asr/data/text_to_text.py

```diff
@@ -0,0 +1,462 @@
+                try:
+                    tts_texts.append(tmp_item["tts_text_normalized"])
+                except KeyError:
+                    tts_texts.append(tmp_item["tts_text"])
+                    need_normalization = True
+
+        if need_normalization:
+            logging.warning("TTS normalization is extremely slow! It is recommended to normalize TTS text")
+
+        if num_skipped_utterances:
+            logging.warning(f"Skipped {num_skipped_utterances} utterances " f"with {num_skipped_words}")
+
+        num_utterances = len(asr_texts)
+        # preprocessing is very costly, if we need only part - remove unnecessary utterances
+        if num_parts > 1:
+            # NB: floor division, full dataset can contain fewer utterances than original, like in tarred dataset
+            num_utterances_part = num_utterances // num_parts
+            start = num_utterances_part * current_part_index
+            end = start + num_utterances_part
+            logging.info(
+                f"Taking part of the dataset: {current_part_index} index, total {num_parts} from {start} to {end}"
+            )
+            asr_texts = asr_texts[start:end]
+            tts_texts = tts_texts[start:end]
+            num_utterances = num_utterances_part
+
+        self.data: List[Dict[str, Any]] = [dict() for _ in range(num_utterances)]
+
+        if tokenizer_workers == 1:
+            logging.warning("Preprocessing large text with tokenizer_workers=1 may be slow with ASR tokenizer")
+            for i, tokenized_text in enumerate(
+                tqdm((self._asr_text_to_tokens(text) for text in asr_texts), total=len(asr_texts))
+            ):
+                self.data[i]["asr_text_tokens"] = tokenized_text
+        else:
+            # Multiprocessing hack: use global variables for every process (not really global in program context)
+            def _init_asr_tokenize_process(tokenizer, bos_id, eos_id):
+                global asr_tokenizer_global, asr_bos_id_global, asr_eos_id_global  # process-global
+                # deepcopy to avoid serialization of parent models
+                asr_tokenizer_global = copy.deepcopy(tokenizer)
+                asr_bos_id_global = copy.deepcopy(bos_id)
+                asr_eos_id_global = copy.deepcopy(eos_id)
+
+            chunksize = 1000  # chunk size for pool map, higher is faster, but less responsive
+            with concurrent.futures.ProcessPoolExecutor(
+                initializer=_init_asr_tokenize_process,
+                initargs=(asr_tokenizer, self.asr_bos_id, self.asr_eos_id),
+                max_workers=tokenizer_workers,
+            ) as pool:
+                for i, tokenized_text in enumerate(
+                    tqdm(pool.map(_asr_text_to_tokens, asr_texts, chunksize=chunksize), total=len(asr_texts))
+                ):
+                    self.data[i]["asr_text_tokens"] = tokenized_text
+        # force free memory
+        del asr_texts
+        gc.collect()
```

Comment:
Is this necessary?

Reply:
Yes. To avoid huge memory usage, I need to delete original texts from memory.
Unfortunately, Python **doesn't guarantee** that `del` will actually delete an object and release the memory (and also call `__del__` object method), even if we have 0 references to the object. The only way to guarantee it is to call `gc.collect`.

Some sources:
- https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__
- https://stackoverflow.com/a/47978346